Assets/Scripts/Framework/Train/TrainEngine.cs

```diff
@@ -0,0 +1,257 @@
+
+        [SerializeField]
+        private float _maxLowSpeed = 2.5f;
+
+        [SerializeField, Space]
+        private float _brakeDuration = 1.5f;
+        
+        [SerializeField]
+        private float _accelerationDuration = 10f;
+
+        [SerializeField]
+        private Ease _accelerationEase = Ease.Linear;
+
+        /// <summary>
+        /// Current train engine state.
+        /// </summary>
+        public TrainEngineState EngineState 
+            { get; private set; } = TrainEngineState.OnStandby;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentTargetEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train velocity speed.
+        /// </summary>
+        public float CurrentVelocity => _currentSpeed / _trainController.SplineLength;
+
+        /// <summary>
+        /// Determines the chosen track for the next possible rail split.<br></br>
+        /// <br>False = left.</br>
+        /// <br>True = right.</br>
+        /// </summary>
+        public bool CurrentSplitDirection { get; set; }
+
+        /// <summary>
+        /// Invokes when train direction is changed.
+        /// </summary>
+        public Action<bool> OnDirectionChanged;
+
+        /// <summary>
+        /// Invoked when the train speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train target speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnTargetSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train engine state is changed.
+        /// </summary>
+        public Action<TrainEngineState> OnEngineStateChanged;
+
+        private Dictionary<TrainEngineSpeedTypes, float> _speedValues;
+
+        [HideInInspector]
+        [SyncVar(Channel = FishNet.Transporting.Channel.Reliable)]
+        private float _currentSpeed;
+        private float _speedTypesCount;
+
+        private Sequence _speedSequence;
+        private TrainController _trainController;
+
+        private void Awake() => _trainController = GetComponent<TrainController>();
+
+        private void Start()
+        {
+            _speedTypesCount = Enum.GetValues(typeof(TrainEngineSpeedTypes)).Length - 1;
+
+            // Initialize speed values dictionary.
+            _speedValues = new Dictionary<TrainEngineSpeedTypes, float>()
+            {
+                {TrainEngineSpeedTypes.High, _maxHighSpeed },
+                {TrainEngineSpeedTypes.Low, _maxLowSpeed },
+                {TrainEngineSpeedTypes.Medium, _maxMediumSpeed },
```

Comment:
I would change this order a bit.
```suggestion
                {TrainEngineSpeedTypes.High, _maxHighSpeed },
                {TrainEngineSpeedTypes.Medium, _maxMediumSpeed },
                {TrainEngineSpeedTypes.Low, _maxLowSpeed },
```